File: reference_engine/degree_hours.py
"""Weighted degree-hour calculations and annual-load disaggregation."""

from __future__ import annotations

import math
from collections import defaultdict
from collections.abc import Sequence

from .models import ClimateRecord


class DegreeHourError(ValueError):
    """Raised when climate weights and annual loads are inconsistent."""


def heating_degree_hour(air_temp_c: float, balance_temperature_c: float) -> float:
    return max(0.0, float(balance_temperature_c) - float(air_temp_c))


def cooling_degree_hour(air_temp_c: float, balance_temperature_c: float) -> float:
    return max(0.0, float(air_temp_c) - float(balance_temperature_c))


def validate_climate_records(
    records: Sequence[ClimateRecord],
    expected_annual_weight_hours: float,
    absolute_tolerance: float,
) -> tuple[str, ...]:
    if not records:
        raise DegreeHourError("At least one climate record is required")
    tolerance = abs(float(absolute_tolerance))
    total_weight = 0.0
    warnings: list[str] = []
    for index, record in enumerate(records):
        if not math.isfinite(record.air_temp_c):
            raise DegreeHourError(f"Record {index} has a non-finite air temperature")
        if not math.isfinite(record.weight_hours) or record.weight_hours <= tolerance:
            raise DegreeHourError(f"Record {index} must have positive finite weight_hours")
        if record.month < 1 or record.month > 12:
            raise DegreeHourError(f"Record {index} has an invalid month")
        total_weight += record.weight_hours
    if abs(total_weight - expected_annual_weight_hours) > tolerance:
        warnings.append(
            f"Climate weights sum to {total_weight:g} h, not the configured "
            f"{expected_annual_weight_hours:g} h."
        )
    return tuple(warnings)


def weighted_degree_hours(
    records: Sequence[ClimateRecord],
    heating_balance_temperature_c: float,
    cooling_balance_temperature_c: float,
) -> dict[str, float]:
    heating = sum(
        heating_degree_hour(record.air_temp_c, heating_balance_temperature_c)
        * record.weight_hours
        for record in records
    )
    cooling = sum(
        cooling_degree_hour(record.air_temp_c, cooling_balance_temperature_c)
        * record.weight_hours
        for record in records
    )
    return {"heating": heating, "cooling": cooling}


def allocate_annual_load(
    records: Sequence[ClimateRecord],
    annual_load_kwh: float,
    balance_temperature_c: float,
    mode: str,
    zero_degree_hour_policy: str,
    absolute_tolerance: float,
) -> list[float]:
    load = float(annual_load_kwh)
    tolerance = abs(float(absolute_tolerance))
    if not math.isfinite(load) or load < -tolerance:
        raise DegreeHourError("annual_load_kwh must be finite and non-negative")
    if mode == "heating":
        degrees = [
            heating_degree_hour(record.air_temp_c, balance_temperature_c)
            for record in records
        ]
    elif mode == "cooling":
        degrees = [
            cooling_degree_hour(record.air_temp_c, balance_temperature_c)
            for record in records
        ]
    else:
        raise DegreeHourError("mode must be 'heating' or 'cooling'")
    weighted = [degree * record.weight_hours for degree, record in zip(degrees, records)]
    denominator = sum(weighted)
    if load <= tolerance:
        return [0.0 for _ in records]
    if denominator > tolerance:
        return [load * contribution / denominator for contribution in weighted]
    if zero_degree_hour_policy == "uniform":
        total_weight = sum(record.weight_hours for record in records)
        if total_weight <= tolerance:
            raise DegreeHourError("Cannot uniformly allocate load with zero total weight")
        return [load * record.weight_hours / total_weight for record in records]
    if zero_degree_hour_policy == "discard_with_warning":
        return [0.0 for _ in records]
    if zero_degree_hour_policy == "error":
        raise DegreeHourError(
            f"Annual {mode} load is positive but weighted {mode} degree-hours are zero"
        )
    raise DegreeHourError(
        "zero_degree_hour_policy must be 'error', 'uniform', or "
        "'discard_with_warning' in the reference engine"
    )


def aggregate_values(
    records: Sequence[ClimateRecord],
    values: Sequence[float],
    selected_period_flags: Sequence[bool] | None = None,
) -> dict[str, object]:
    if len(records) != len(values):
        raise DegreeHourError("records and values must have the same length")
    flags = (
        [False for _ in records]
        if selected_period_flags is None
        else list(selected_period_flags)
    )
    if len(records) != len(flags):
        raise DegreeHourError(
            "records and selected_period_flags must have the same length"
        )
    monthly: dict[int, float] = defaultdict(float)
    monthly_selected: dict[int, float] = defaultdict(float)
    annual = 0.0
    selected = 0.0
    for record, value, is_selected in zip(records, values, flags):
        amount = float(value)
        annual += amount
        monthly[record.month] += amount
        if is_selected:
            selected += amount
            monthly_selected[record.month] += amount
    return {
        "annual": annual,
        "selected_period": selected,
        "monthly": {str(month): monthly[month] for month in range(1, 13)},
        "monthly_selected_period": {
            str(month): monthly_selected[month] for month in range(1, 13)
        },
    }

File: src/engine/degree-hours.ts
import { DegreeHourError } from "./errors";
import type { ClimateRecord, ValueAggregate } from "./types";

export function heatingDegreeHour(airTempC: number, balanceTemperatureC: number): number {
  return Math.max(0, balanceTemperatureC - airTempC);
}

export function coolingDegreeHour(airTempC: number, balanceTemperatureC: number): number {
  return Math.max(0, airTempC - balanceTemperatureC);
}

export function validateClimateRecords(
  records: readonly ClimateRecord[],
  expectedAnnualWeightHours: number,
  absoluteTolerance: number,
): string[] {
  if (records.length === 0) throw new DegreeHourError("At least one climate record is required");
  const tolerance = Math.abs(absoluteTolerance);
  let totalWeight = 0;
  records.forEach((record, index) => {
    if (!Number.isFinite(record.airTempC)) {
      throw new DegreeHourError(`Record ${index} has a non-finite air temperature`);
    }
    if (!Number.isFinite(record.weightHours) || record.weightHours <= tolerance) {
      throw new DegreeHourError(`Record ${index} must have positive finite weightHours`);
    }
    if (record.month < 1 || record.month > 12) {
      throw new DegreeHourError(`Record ${index} has an invalid month`);
    }
    totalWeight += record.weightHours;
  });
  return Math.abs(totalWeight - expectedAnnualWeightHours) > tolerance
    ? [
        `Climate weights sum to ${totalWeight} h, not the configured ${expectedAnnualWeightHours} h.`,
      ]
    : [];
}

export function weightedDegreeHours(
  records: readonly ClimateRecord[],
  heatingBalanceTemperatureC: number,
  coolingBalanceTemperatureC: number,
): { heating: number; cooling: number } {
  let heating = 0;
  let cooling = 0;
  for (const record of records) {
    heating +=
      heatingDegreeHour(record.airTempC, heatingBalanceTemperatureC) * record.weightHours;
    cooling +=
      coolingDegreeHour(record.airTempC, coolingBalanceTemperatureC) * record.weightHours;
  }
  return { heating, cooling };
}

export function allocateAnnualLoad(
  records: readonly ClimateRecord[],
  annualLoadKwh: number,
  balanceTemperatureC: number,
  mode: "heating" | "cooling",
  zeroDegreeHourPolicy: "error" | "uniform" | "discard_with_warning",
  absoluteTolerance: number,
): number[] {
  const tolerance = Math.abs(absoluteTolerance);
  if (!Number.isFinite(annualLoadKwh) || annualLoadKwh < -tolerance) {
    throw new DegreeHourError("annualLoadKwh must be finite and non-negative");
  }
  const degrees = records.map((record) =>
    mode === "heating"
      ? heatingDegreeHour(record.airTempC, balanceTemperatureC)
      : coolingDegreeHour(record.airTempC, balanceTemperatureC),
  );
  const weighted = degrees.map((degree, index) => degree * records[index]!.weightHours);
  const denominator = weighted.reduce((sum, value) => sum + value, 0);
  if (annualLoadKwh <= tolerance) return records.map(() => 0);
  if (denominator > tolerance) {
    return weighted.map((contribution) => (annualLoadKwh * contribution) / denominator);
  }
  if (zeroDegreeHourPolicy === "uniform") {
    const totalWeight = records.reduce((sum, record) => sum + record.weightHours, 0);
    if (totalWeight <= tolerance) {
      throw new DegreeHourError("Cannot uniformly allocate load with zero total weight");
    }
    return records.map((record) => (annualLoadKwh * record.weightHours) / totalWeight);
  }
  if (zeroDegreeHourPolicy === "discard_with_warning") return records.map(() => 0);
  throw new DegreeHourError(
    `Annual ${mode} load is positive but weighted ${mode} degree-hours are zero`,
  );
}

export function aggregateValues(
  records: readonly ClimateRecord[],
  values: readonly number[],
  selectedPeriodFlags?: readonly boolean[],
): ValueAggregate {
  if (records.length !== values.length) {
    throw new DegreeHourError("records and values must have the same length");
  }
  const flags = selectedPeriodFlags ?? records.map(() => false);
  if (flags.length !== records.length) {
    throw new DegreeHourError("records and selectedPeriodFlags must have the same length");
  }
  const monthly = Object.fromEntries(Array.from({ length: 12 }, (_, index) => [String(index + 1), 0]));
  const monthlySelectedPeriod = Object.fromEntries(
    Array.from({ length: 12 }, (_, index) => [String(index + 1), 0]),
  );
  let annual = 0;
  let selectedPeriod = 0;
  records.forEach((record, index) => {
    const amount = values[index]!;
    annual += amount;
    const month = String(record.month);
    monthly[month] = (monthly[month] ?? 0) + amount;
    if (flags[index]) {
      selectedPeriod += amount;
      monthlySelectedPeriod[month] = (monthlySelectedPeriod[month] ?? 0) + amount;
    }
  });
  return { annual, selectedPeriod, monthly, monthlySelectedPeriod };
}
